api: add hardcoded versioning support

GRISHNOV opened the pull request:
Added the _VERSION variable to the exported table. Is part of the task [1].

Khatskevich commented on graphql/accessor_general.lua:
```diff
@@ -230,9 +248,12 @@ end
+--- `value_list`
+---
+--- @treturn table `value_list` (optional) is values list from the `filter`
+--- argument ordered in the such way that can be passed to the found index (has
```
in such way that it can be passed

Totktonada replied:
Fixed.

Khatskevich commented on graphql/accessor_general.lua:
```diff
@@ -230,9 +248,12 @@ end
---- @treturn table `value_list` is values list from the `filter` argument
---- ordered in the such way that can be passed to the found index (has some
---- meaning only when `index_name ~= nil`)
+--- @treturn table `new_filter` is the filter value w/o values extracted into
```
I suggest name `brute_filter` for more semantics

Totktonada replied:
I generally agree that common names are bad for domain-specific things. I’ll consider `post_filter` names or smth like so when there will less pending PRs.

There are related questions / points:

1. `post_filter` (postprocessing filter) name will loose some of its meaning once we’ll support pushdown filtering to sotrage nodes.
2. There are two other things of the same (at least for now) structure: pivot.filter (slow case of offset) and determinator of union-type connections (#8).

Khatskevich commented on graphql/accessor_general.lua:
```diff
@@ -402,6 +441,34 @@ local function build_lookup_index_name(indexes)
 
+--- Build `parts_tree` to use in @{get_index_name} for lookup best matching
+--- index.
+---
```
It required from me some time to get the return value structure.
I suppose it would be great if you draw small schematic image here. e.g.
```
transformation example
i1: f1 f2 f3 f4
i2: f1 f2 f4
become
f1--- f2 -- f3 -- f4 --> i1
         `-- f4 --> i2
```

Totktonada replied:
Done.

Khatskevich commented on graphql/accessor_general.lua:
```diff
@@ -121,14 +121,42 @@ local function filter_names_fingerprint(filter)
+--- @tparam table filter map of key-value to filter objects against
+---
+--- @treturn string `index_name` or `nil` is found index
+---
```
I want some information about its running time here.
I tried to calculate it and I suppose it is something like O(SUM(key_parts)*SIZE(filter)) in worst case which is good for most use cases.

Totktonada replied:
Added.

Khatskevich commented on test/testdata/compound_index_testdata.lua:
```diff
@@ -231,6 +231,13 @@ function compound_index_testdata.run_queries(gql_wrapper)
+        local variables_1_3 = {user_str = 'user_str_b'}
+        local result = gql_query_1:execute(variables_1_3)
+        results = results .. print_and_return(
+            ('RESULT\n%s'):format(yaml.encode(result)))
```
Those tests do not check if new functions work correctly(
Tests on added feature needed.

Totktonada replied:
1. Done in the latter commit.
2. We check here that functionality was not broken after the change. Index lookup check depends on metainfo (#71), will be tracked by #72.

Totktonada replied:
Are you about that here?

Khatskevich commented on test/testdata/compound_index_testdata.lua:
```diff
@@ -14,6 +14,11 @@ local function print_and_return(...)
     return table.concat({...}, ' ') .. '\n'
 end
 
+local function format_result(name, query, variables, result)
```
This feature makes review process much easier)
However to make the previous commit readable, could you please move this commit on top?)

Khatskevich replied:
Actually, I do not want you to waste your time. By now it is not important anymore.

Totktonada commented on test/shard_servers/master.lua:
```diff
@@ -1,5 +1,7 @@
 #!/usr/bin/env tarantool
```
Fixed.

Totktonada commented on test/shard_servers/master.lua:
```diff
@@ -1,5 +1,7 @@
 #!/usr/bin/env tarantool
 
+shard_initialized = false
```
Thanks! Fixed.

Totktonada commented on test/testdata/compound_index_testdata.lua:
```diff
@@ -231,6 +231,13 @@ function compound_index_testdata.run_queries(gql_wrapper)
     end)
 
+    utils.show_trace(function()
+        local variables_1_3 = {user_str = 'user_str_b'}
```
Will be done in the scope of non-prio1 issues, tracked by #72.

Totktonada commented on graphql/accessor_general.lua:
```diff
@@ -121,14 +121,42 @@ local function filter_names_fingerprint(filter)
+                get_best_matched_index(successor_node, new_filter)
+            if branch_index_name ~= nil and branch_len > max_len then
+                index_name = branch_index_name
+                max_len = branch_len
```
Nope. It needed for comparing branches. I’ll rename it to `max_branch_len` to avoid confusion.

Totktonada replied:
Also I’ll change description of `len` return argument as (matched parts + 1) and will describe the reason why.

Totktonada replied:
Moved branch_len = branch_len + 1 to here.

Khatskevich commented on graphql/accessor_general.lua:
```diff
@@ -121,14 +121,86 @@ local function filter_names_fingerprint(filter)
+            local new_filter = table.copy(filter)
+            new_filter[k] = nil
+            local branch_index_name, branch_len =
+                get_best_matched_index(successor_node, new_filter)
```
Please, move + 1 here

Totktonada replied:
Done. Still don’t see why it is better, but changed because of lack of any opinion about that.

Khatskevich commented on graphql/accessor_general.lua:
```diff
@@ -121,14 +121,86 @@ local function filter_names_fingerprint(filter)
+--- Complexity
+--- ----------
+---
+--- In short: O(N^2 * COUNT(index parts)).
```
Provide here information, that n= len(filter)

Totktonada replied:
Done.

Khatskevich commented on graphql/accessor_general.lua:
```diff
@@ -402,6 +485,64 @@ local function build_lookup_index_name(indexes)
+---          \             \
+---           \             + --> abc
+---            \
+---             + ------> abc --> efg --> hij
```
Please draw also index names as leaf nodes

Totktonada replied:
Leaf nodes are actually 'abc', 'hij' and so on, but I added index names.

Totktonada commented on graphql/accessor_general.lua:
```diff
@@ -121,14 +121,86 @@ local function filter_names_fingerprint(filter)
+--- Future optimizations
+--- --------------------
+---
+--- * replace table.copy() with something more light (maybe 'closed set' of
```
I suspect that it is not so easy, because of pairs() invalidation.

Totktonada commented on graphql/accessor_general.lua:
```diff
@@ -121,14 +121,86 @@ local function filter_names_fingerprint(filter)
+--- (M * K)). Nodes count (M * K) can be limited upside as count of index
+--- parts in all indexes, so we have O(N^2 * COUNT(index parts)).
+---
+--- We can consider worst case scenario when any node has any of filter keys as
```
Different views on the complexity problem: dependency of filter size in context of fixed indexes or in context of worst-case indexes. Maybe it is something one should not consider.

GRISHNOV committed:
api: add hardcoded versioning support

Added the _VERSION variable to the exported table.
Is part of the task [1].

1. github.com/tarantool/roadmap-internal/issues/204

DifferentialOrange commented on graphql/init.lua:
```diff
@@ -1 +1 @@
-return {VERSION = 'unknown'}
+return {_VERSION = require('graphql.version')}
```
CI code for automatic version generation still exist. You need to remove it first.

See https://github.com/tarantool/graphql/commit/0caab752868826f4367dbcf0cf71ecf6089a8f60

GRISHNOV replied:
Deleted

DifferentialOrange commented on graphql/init.lua:
```diff
@@ -1 +1 @@
-return {VERSION = 'unknown'}
+return {_VERSION = require('graphql.version')}
```
Let's use the following approach: if module already has a version handle, we will not remove it. (It would be the best to add an issue about `.VERSION` deprecation, but let's do it later.) So let's preserve two variables for a time (but let their values be equal -- let the old version use the new approach).

GRISHNOV replied:
Done
